docs: edit and update DI providers (#39403)
angular/angular: aio/content/guide/dependency-injection-providers.md

# Dependency providers

> You can use dependency providers to get your services where you need them.
> By configuring providers, you can access dependencies in components, directives, pipes, and other services.

Reviewer: ```suggestion
By configuring providers, you can make services available to the parts of your application that need them.
```

The following example is the class provider syntax for providing a `Logger` class in the `providers` array.

<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-logger">
</code-example>

Angular expands the `providers` value into a full provider object as follows.

<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-3" >
</code-example>

The expanded provider configuration is an object literal with two properties.

1. The `provide` property holds the [token](guide/dependency-injection#token) that serves as the key for both locating a dependency value and configuring the injector.
2. The second property is a provider definition object, which tells the injector how to create the dependency value.
The provider-definition key can be `useClass`, as in the example.
It can also be `useExisting`, `useValue`, or `useFactory`.
Each of these keys provides a different type of dependency, as discussed below.

{@a class-provider}

## Configuring the injector to use alternative class providers

> To configure the injector to return a different class that provides the same service, you can use `useClass`.

Reviewer: ```suggestion
To configure the injector to return a different class that provides the same service, you can use the `useClass` property.
```

In this example, the injector returns a `BetterLogger` instance when using the `Logger` token.

<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-4" >
</code-example>

{@a class-provider-dependencies}

### Configuring class providers with dependencies

If the alternative class providers have their own dependencies, specify both providers in the `providers` metadata option of the parent module or component.

<code-example path="dependency-injection/src/app/providers.component.ts" region="providers-5"></code-example>

> In this example, `EvenBetterLogger` display the user name in the log message.

Reviewer: ```suggestion
In this example, `EvenBetterLogger` displays the user name in the log message.
```

This logger gets the user from an injected `UserService` instance.

<code-example path="dependency-injection/src/app/providers.component.ts" region="EvenBetterLogger"></code-example>

The injector needs providers for both this new logging service and its dependent `UserService`.

{@a aliased-class-providers}

### Aliasing class providers

> To alias a class provider, specify the alias and the class provider in the `providers` array with `useExisting`.

Reviewer: ```suggestion
To alias a class provider, specify the alias and the class provider in the `providers` array with the `useExisting` property.
```

Now you can add a parent provider to your components that's easier to read and understand.

<code-example path="dependency-injection-in-action/src/app/parent-finder.component.ts" region="alice-providers" header="dependency-injection-in-action/src/app/parent-finder.component.ts"></code-example>

### Aliasing multiple class interfaces

To alias multiple parent types, each with its own class interface token, configure `provideParent()` to accept more arguments.

Here's a revised version that defaults to `parent` but also accepts an optional second parameter for a different parent class interface.

<code-example path="dependency-injection-in-action/src/app/parent-finder.component.ts" region="provide-parent" header="dependency-injection-in-action/src/app/parent-finder.component.ts"></code-example>

Next, to use `provideParent()` with a different parent type, provide a second argument, here `DifferentParent`.

<code-example path="dependency-injection-in-action/src/app/parent-finder.component.ts" region="beth-providers" header="dependency-injection-in-action/src/app/parent-finder.component.ts"></code-example>

{@a value-provider}

## Injecting an object

> To inject an object, configure the injector with the `useValue` option.

Reviewer: Just a note: let's be consistent with property or option. Our API says `useValue` and `useClass` are properties, so maybe that's the right way to go?

Author: Agreed. `option` appears frequently so I'll switch them as we go.
